crates/sui-data-ingestion-core/src/reader.rs

```diff
@@ -61,31 +65,81 @@ impl CheckpointReader {
+        store: &dyn ObjectStore,
+        checkpoint_number: CheckpointSequenceNumber,
+    ) -> Result<CheckpointData> {
+        let path = Path::from(format!("{}.chk", checkpoint_number));
+        let response = store.get(&path).await?;
+        let bytes = response.bytes().await?;
+        Blob::from_bytes::<CheckpointData>(&bytes)
+    }
+
+    fn start_remote_fetcher(&mut self) -> mpsc::Receiver<Result<CheckpointData>> {
+        let batch_size = self.remote_read_batch_size;
+        let start_checkpoint = self.current_checkpoint_number;
+        let (sender, receiver) = mpsc::channel(batch_size);
+        let url = self
+            .remote_store_url
+            .clone()
+            .expect("remote store url must be set");
+        let store = if self.remote_store_options.is_empty() {
+            let retry_config = RetryConfig {
+                max_retries: 0,
+                retry_timeout: Duration::from_secs(10),
+                ..Default::default()
+            };
+            let http_store = object_store::http::HttpBuilder::new()
+                .with_url(url)
+                .with_retry(retry_config)
+                .build()
+                .expect("failed to parse remote store config");
+            Box::new(http_store)
+        } else {
+            parse_url_opts(
+                &Url::parse(&url).expect("failed to parse remote store url"),
+                self.remote_store_options.clone(),
+            )
+            .expect("failed to parse remote store config")
+            .0
+        };
+
+        spawn_monitored_task!(async move {
+            let mut checkpoint_stream = (start_checkpoint..u64::MAX)
+                .map(|checkpoint_number| Self::remote_fetch_checkpoint(&store, checkpoint_number))
+                .pipe(futures::stream::iter)
+                .buffered(batch_size);
+
+            while let Some(checkpoint) = checkpoint_stream.next().await {
+                if sender.send(checkpoint).await.is_err() {
                     break;
                 }
-                let checkpoint = Blob::from_bytes::<CheckpointData>(&bytes.unwrap())?;
-                checkpoints.push(checkpoint);
             }
+        });
+        receiver
+    }
+
+    fn remote_fetch(&mut self) -> Vec<CheckpointData> {
+        let mut checkpoints = vec![];
+        if self.remote_fetcher_receiver.is_none() {
+            self.remote_fetcher_receiver = Some(self.start_remote_fetcher());
         }
-        Ok(checkpoints)
+        while !self.exceeds_capacity(self.current_checkpoint_number + checkpoints.len() as u64) {
+            match self.remote_fetcher_receiver.as_mut().unwrap().try_recv() {
+                Ok(Ok(checkpoint)) => checkpoints.push(checkpoint),
+                Ok(Err(err)) => {
+                    info!("remote reader transient error {:?}", err);
+                    self.remote_fetcher_receiver = None;
+                    return checkpoints;
+                }
+                Err(_) => break,
```

Comment:
There seems to be two different error conditions that probably need to be handled differently.

The first `Ok(Err(err))` seems to be a fetch that errored. `Err(_)` has [two cases](https://docs.rs/tokio/latest/tokio/sync/mpsc/error/enum.TryRecvError.html), one where the channel is empty and we just need to wait some time till the other side can fill it, and disconnected where the channel will never have anymore data.

Reply:
yeah, Ok(Err(err)) is the fetch that errored. It's ok to just log it and move on, `self.remote_fetcher_receiver` is getting canceled further in the block and will get respawned on the next tick.
Other break clause was written specifically for Empty clause(i.e. move on until next so fetcher may to accumulate more tasks).
So yeah, probably `Disconnected` is not handled properly, although I'm not sure how it will get triggered without an error in corresponding task. But will add explicit handler